nemo/utils/exp_manager.py

```diff
@@ -470,10 +470,18 @@ def check_resume(
             for fold in other_run_dirs:
                 if fold.is_dir():
                     run_count += 1
+            logging.info(f"Backing up run_{run_count}")
             new_run_dir = Path(Path(log_dir) / f"run_{run_count}")
             new_run_dir.mkdir()
             for _file in files_to_move:
                 move(str(_file), str(new_run_dir))
+            time.sleep(10)
+    else:
+        files_to_move = [child for child in Path(log_dir).iterdir() if child.is_file()]
+        while len(files_to_move) > 0:
+            logging.info(f"Waiting for global_rank_zero() to finish back up of previous run")
+            files_to_move = [child for child in Path(log_dir).iterdir() if child.is_file()]
+            time.sleep(5)
 
```

Comment:
Instead of busy waiting, check if in distributed torch mode, then use torch.distributed.barrier()

Reply:
That will unfortunately not work, at that point in code `torch.distributed` is not yet initialised (I have just checked), hence `torch.distributed.barrier()` cannot be used. 

This should make no difference, but just for clarity; I am running using Slurm 21.08.5, which was deployed via [nvidia/deepos](https://github.com/NVIDIA/deepops), and support for containers is provided by [nvidia/enroot](https://github.com/NVIDIA/enroot) and [nvidia/pyxis](https://github.com/NVIDIA/pyxis). Logs contain a line `Multiprocessing is handled by SLURM.` At least in this setting it seems that `torch.distributed` gets initialised much later, indeed I mentioned in #4300 that it is not initialised even when NLP/PC DataLoaders get initialised. This may be true for other nemo/collections that make use of `torch.distributed.barrier()` as well. 

At the moment I don’t have a better solution than busy waiting. In fact I’ve added additional sleep cycles to ensure all loops exit before any files are created.

Reply:
@titu1994 any comments? Is there anything else that could be used as a barrier/semaphore?

Reply:
@titu1994 reviewing PTL code, one can see, that `Trainer.strategy.setup_environment()` is called only after calling `Trainer.fit()`, `Trainer.test()` or `Trainer.predict()`, see https://github.com/Lightning-AI/lightning/blob/8d14554383632f4cdff337b3d8d1b226cabdd1d0/src/pytorch_lightning/trainer/trainer.py#L1114-L1136

`Trainer.strategy.setup_environment()` will call `Strategy.setup_distributed()`, which will call `Strategy.init_dist_connection()`, which finaly does `torch.distributed.init_process_group()`, therefore initialises `torch.distribured` and makes it available, and also makes `barrier()` a possibility. 

Based on https://github.com/Lightning-AI/lightning/blob/8d14554383632f4cdff337b3d8d1b226cabdd1d0/src/pytorch_lightning/strategies/ddp.py#L307-L313 I would also assume that it is better to call `trainer.strategy.barrier()` than `torch.distributed.barier()` directly.

In other words, the only way of making `torch.distributed` available earlier is to add the following two lines immediately after trainer setup in all `nemo/examples`
```python3
    if trainer.strategy.is_distributed:
        trainer.strategy.setup_environment()
```
with all synchronisations can be handled via `trainer.strategy.barrier()`.

If you prefer this solution, let me know and I will update the PR, but it will touch a lot of nemo code, i.e. all examples and all calls to `torch.distributed.barrier()`. I'm no expert in PTL, so if you have someone on the inside, do double check.